### ISSUE TITLE
Back button does not work on web based CWTS w/ context=fx_desktop_v2

### ISSUE THREAD

**Issue by shane-tomlinson** [mozilla/fxa-content-server#3329]
STR

1. Load up a fresh profile in Nightly.
1. Open "about:preferences#sync"
1. Click "create account"
1. Fill out and submit the signup form.
1. Click the "back" button. Nothing happens.
1. Click the "back" button again. FxA is replaced with "about:preferences#sync"

![back-button-broken-fx-desktop-v2](https://cloud.githubusercontent.com/assets/848085/11842955/c3a3a398-a3fe-11e5-8a0b-f03b8fe897cb.gif)

ref #3330

**Comment by vladikoff** [comment]
Oh weird, we should add a functional testing to that

**Comment by vladikoff** [comment]
Hm strange, running this via the tests the button works fine. Something strange going on

![wat2](https://cloud.githubusercontent.com/assets/128755/11846760/fe6b3f52-a3e7-11e5-9b45-843e211efede.gif)

**Comment by vladikoff** [comment]
Looks like some about:accounts business to blame

![](http://v14d.com/i/5671914ed8b6b.jpg)

**Comment by shane-tomlinson** [comment]
Everything works fine locally when loading `about:accounts`.

**Comment by shane-tomlinson** [comment]
Ahha! Successful reproduction locally:

1. Set up Firefox to use a local FxA stack.
1. Load [about:accounts?action=signup&entrypoint=menupanel](about:accounts?action=signup&entrypoint=menupanel)
1. Sign up for a new account.
1. On CWTS, click "back".
1. Wah wah.

The `?action=signup&entrypoint=menupanel` portion of the URL is the important part, without it, the "back" button works as expected.

**Comment by shane-tomlinson** [comment]
I'm having a hard time understanding what's going on.

With 42.0a2 (2015-08-14), everything works fine in `about:accounts?action=signup&entrypoint=menupanel` and `about:accounts`. Using Beta (43.0) or Nightly 46.0a1 (2015-12-16), I can go forward from signup to CWTS, but not "back", with the same error @vladikoff shows in https://github.com/mozilla/fxa-content-server/issues/3329#issuecomment-165165562.

With this diff applied, everything works as expected:

```diff
diff --git a/app/scripts/lib/app-start.js b/app/scripts/lib/app-start.js
index de8f884..efc72fb 100644
--- a/app/scripts/lib/app-start.js
+++ b/app/scripts/lib/app-start.js
@@ -585,7 +585,7 @@ define(function (require, exports, module) {
     allResourcesReady: function () {
       // The IFrame cannot use pushState or else a page transition
       // would cause the parent frame to redirect.
-      var usePushState = ! this._isInAnIframe();
+      var usePushState = false;//! this._isInAnIframe();
 
       if (! usePushState) {
         // If pushState cannot be used, Backbone falls back to using
```

After adding a lot of console messages, I think there are multiple problems occurring. First, a screen capture:

![screen shot 2015-12-17 at 12 15 44](https://cloud.githubusercontent.com/assets/848085/11869466/f5b40896-a4b7-11e5-9d4a-5740831f0a44.png)

The first orange entry is going from `signup`=>`choose_what_to_sync`.
The second orange entry is the interesting one: an attempt is made to go from `about:blank` to `settings`.

A `SecurityError` is immediately thrown.

[Look at this portion of the WhatWG HTML spec](https://html.spec.whatwg.org/multipage/browsers.html#dom-history-pushstate):

----------------------------------

* If this History object is associated with a Document that is not fully active, throw a SecurityError.
* Let cloned data be a structured clone of the specified data. If this throws an exception, 
then rethrow that exception and abort these steps.
* If the third argument is not null, run these substeps:
  1. Resolve the value of the third argument, relative to the API base URL specified by the entry settings object.
  1. If that fails, throw a SecurityError exception and abort these steps.
  1. Compare the resulting parsed URL to the result of applying the URL parser algorithm to the document's address. If any component of these two URLs differ other than the path, query, and fragment components, then throw a SecurityError exception and abort these steps.
  1. If the origin of the resulting absolute URL is not the same as the origin of the responsible document specified by the entry settings object, and either the path or query components of the two parsed URLs compared in the previous step differ, throw a SecurityError exception and abort these steps. (This prevents sandboxed content from spoofing other pages on the same origin.)
  1. Let new URL be the resulting absolute URL.
* For the purposes of the comparisons in the above substeps, the path and query components can only be the same if the scheme component of both parsed URLs are relative schemes.

---------------------------------------

It seems like with signup->CWTS->back `window.history.back` is going back to `about:blank`. This for whatever reason is taken to be `/`. The `/` [route handler kicks in](https://github.com/mozilla/fxa-content-server/blob/eccb4366036952ca68ecae10963a847d618d0316/app/scripts/lib/router.js#L140-L144), a sessionToken exists for the account so the user is kicked to `/settings`. `/settings` is a relative path, so it's host is resolved based off of the current address (about:accounts). This fails, so the SecurityError from `ii` kicks in. Either that, or the `SecurityError` from `iii` kicks in.

It's like no initial history object is ever created.

AGGGH! I wonder if this has anything to do with it!

https://hg.mozilla.org/mozilla-central/annotate/0babaa3edcf908c393b68a3dc2d1c2a2450c31ed/browser/base/content/aboutaccounts/aboutaccounts.js#l133

The code that opens the window:
```js
    webNav.loadURI(url, Ci.nsIWebNavigation.LOAD_FLAGS_BYPASS_HISTORY, null, null, null);
```

The comment just above:
```
    // Set the iframe's location with loadURI/LOAD_FLAGS_BYPASS_HISTORY to
    // avoid having a new history entry being added.
```

No history object is created on startup to avoid bad "back" button behavior!

cc/ @mhammond

**Comment by shane-tomlinson** [comment]
BOOM! The fix:

```js
diff --git a/app/scripts/lib/app-start.js b/app/scripts/lib/app-start.js
index de8f884..ef80b14 100644
--- a/app/scripts/lib/app-start.js
+++ b/app/scripts/lib/app-start.js
@@ -599,6 +599,16 @@ define(function (require, exports, module) {
       // immediately redirected
       var startPage = this._selectStartPage();
       var isSilent = !! startPage;
+
+      if (usePushState) {
+        // There was a change in Firefox where the initial page in
+        // about:accounts does not create a history entry. If the
+        // user goes to choose what to sync and then clicks "back",
+        // there is no history entry to go back to. Add a history
+        // entry just in case it's needed. See #3329
+        this._window.history.pushState({}, document.title, null);
+      }
+
       this._history.start({ pushState: usePushState, silent: isSilent });
       if (startPage) {
         this._router.navigate(startPage);
```

**Comment by shane-tomlinson** [comment]
@mhammond - my proposed fix obliterates your Fx fix.

**Comment by shane-tomlinson** [comment]
Ref https://bugzilla.mozilla.org/show_bug.cgi?id=969474

**Comment by shane-tomlinson** [comment]
Ref https://bugzilla.mozilla.org/show_bug.cgi?id=1233668

**Comment by shane-tomlinson** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=411469398d76

**Comment by vladikoff** [comment]
Looks like this was fixed in FF Nightly 46.0a1 (2015-12-29) 

Awesome!

**Comment by shane-tomlinson** [comment]
If web based CWTS goes out with with Fx 45, we need the referenced patch uplifted or else the "back" button on the CWTS page will be broken.
